File: packages/llm-test-harness-python/llm_test_harness/metrics/json_schema.py
from __future__ import annotations

import json
from typing import Any

from ..types import MetricFn, MetricScore


class JSONSchema(MetricFn):
    def __init__(self, schema: dict[str, Any]) -> None:
        self._schema = schema

    @property
    def name(self) -> str:
        return "JSONSchema"

    def evaluate(self, text: str) -> MetricScore:
        import jsonschema

        try:
            data = json.loads(text)
        except json.JSONDecodeError as e:
            return MetricScore(name=self.name, passed=False, score=0.0, reason=f"Invalid JSON: {e}")

        validator = jsonschema.Draft7Validator(self._schema)
        errors = list(validator.iter_errors(data))
        if errors:
            reason = "; ".join(e.message for e in errors[:3])
            return MetricScore(name=self.name, passed=False, score=0.0, reason=reason)

        return MetricScore(name=self.name, passed=True, score=1.0)

File: packages/llm-test-harness/src/metrics/json-schema.ts
import Ajv from 'ajv'
import type { MetricFn, MetricScore } from '../types.js'

const ajv = new Ajv({ allErrors: true })

export class JSONSchema implements MetricFn {
  readonly name = 'JSONSchema'
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  private validate: ReturnType<typeof ajv.compile>

  constructor(schema: Record<string, unknown>) {
    this.validate = ajv.compile(schema)
  }

  evaluate(text: string): MetricScore {
    let parsed: unknown
    try {
      parsed = JSON.parse(text)
    } catch {
      return { name: this.name, pass: false, score: 0, reason: 'Output is not valid JSON' }
    }

    const valid = this.validate(parsed)
    if (valid) {
      return { name: this.name, pass: true, score: 1 }
    }

    const reason = this.validate.errors
      ?.map((e) => `${e.instancePath || '(root)'} ${e.message}`)
      .join('; ')
    return { name: this.name, pass: false, score: 0, reason }
  }
}
